[qt] OptionsDialog: add prune setting

Sjors opened the pull request:
The default suggested value is 2 GB. Minimum is 1 GB (550 MB rounded up).

When the user toggles this setting, a strong warning appears that undoing requires re-downloading the chain:

<img width="478" alt="schermafbeelding 2018-05-15 om 12 35 24" src="https://user-images.githubusercontent.com/10217/40051858-7939cc20-583c-11e8-9120-327a75376732.png">

Tooltip points out that actual disk usage can be higher. It's a bit vague on the "advanced features", because I'm assuming anyone who needs to use `-rescan` and `-txindex` will read the documentation, and a more detailed text would needlessly confuse everyone else.

<img width="450" alt="schermafbeelding 2018-05-15 om 12 33 51" src="https://user-images.githubusercontent.com/10217/40051791-49d6156a-583c-11e8-97b9-7de6dfd8c481.png">

The UI uses gigabytes for readability and easy of use. There is also no manual pruning UI (`prune=1`). The user will have to use `bitcoin.conf` for those things.

Fixes #6461. When combined with #13029 the user, after pruning their node, can safely reset settings and/or use bitcoind without having to edit `bitcoin.conf`. However I don't think that's an essential prerequisite.

randolf reviewed (approved):
Adding this feature to the GUI is a great idea.

laanwj commented:
Concept ACK, will test and review. THanks a lot for adding this.

jonasschnelli commented:
Nice! Concept ACK.
Thanks for adding this
One thought: why not adding this (additionally) to the intro screen (including showing the approximate disk usage reduction)?

laanwj commented:
> The default suggested value is 20 GB; ~1 month of worst case data. Minimum is 1 GB (550 MB rounded up).

What would be the motivation to not to default to the minimum?

Sjors commented:
> why not adding this (additionally) to the intro screen (including showing the approximate disk usage reduction)?

That's what I had in mind as well, but I figured it can go in a separate PR.

> What would be the motivation to not to default to the minimum?

To be on the safe side in case of a very large reorg, e.g. if some contentious user activated soft fork gets majority cumulative difficulty after a few weeks. If people think it's overkill, I'm happy to set it to 1 GB. If I remember correctly the limit of 550 MB was decided before SegWit, so maybe 2 GB would be a better default?

laanwj commented:
> To be on the safe side in case of a very large reorg, e.g. if some contentious user activated soft fork gets majority cumulative difficulty after a few weeks. If people think it's overkill, I'm happy to set it to 1 GB. If I remember correctly the limit of 550 MB was decided before SegWit, so maybe 2 GB would be a better default?

Oh, right, good point. From what I remember, the minimum was meant to be 550 blocks, reserving 1MiB per block, so 550MiB. This is no longer a good assumption with segwit blocks.

Also: `NODE_NETWORK_LIMITED` (BIP159) requires 288 blocks.

laanwj commented:
@jonasschnelli can you trigger a build for this PR? Would be useful to have so people can test, I think.

jonasschnelli commented:
Gitian build: https://bitcoin.jonasschnelli.ch/build/584

luke-jr commented:
This adds a new QSettings prune option, distinct from the conf file. But the option cannot be set separately for GUI vs bitcoind, so this makes no sense...

harding commented:
An additional suggestion: if the tooltip isn't already overfull, it might be worth mentioning that the software remains a full node even after limiting block storage, as I find it's a common misconception for people to think that pruned nodes aren't fully validating nodes.  Perhaps saying something like:

> Disables some advanced features but all blocks will still be fully validated. Reverting this setting requires re-downloading the entire blockchain.  Actual disk usage may be somewhat higher.

(5 extra words; 31 extra characters)

mess110 commented:
Tested ACK https://github.com/bitcoin/bitcoin/pull/13043/commits/a47549348afd96aa902babfdacae0160afcc015e

Should the word `prune` appear in the label? Maybe `Prune block storage to`

luke-jr reviewed (changes requested):
Needs to not add a QSettings prune option

Sjors commented:
@luke-jr wrote:
> This adds a new QSettings prune option, distinct from the conf file. But the option cannot be set separately for GUI vs bitcoind, so this makes no sense...

That is what I was referring to in the description:

> When combined with #13029 the user, after pruning their node, can safely reset settings and/or use bitcoind without having to edit bitcoin.conf.

It's why I created #12833 (let QT write to config file, based on your #11082) and alternative #13029 (interpret absence of `prune=` setting in a way that using QSettings for this isn't a big deal).

@harding @mess110 I'll look into the tooltip range and click area. It should probably be broader. Will also look at your suggested text improvement.

@jonasschnelli thanks for the build

Sjors commented on src/qt/optionsdialog.cpp:
```diff
@@ -36,8 +36,17 @@ OptionsDialog::OptionsDialog(QWidget *parent, bool enableWallet) :
     /* Main elements init */
     ui->databaseCache->setMinimum(nMinDbCache);
     ui->databaseCache->setMaximum(nMaxDbCache);
+    static const uint64_t GB = 1024 * 1024 * 1024;
```
I'll rename the variable to GiB, to be consistent with `prune` which also uses `MiB`.

Sjors committed:
[qt] OptionsDialog: add prune setting

Sjors commented:
Rebased, and:

* reduced default to 2 GiB
* fixed tooltip range and click area
* included the suggested tooltip text improvement
* replaced the word "limit" with "prune"
* added `alt+b` keyboard shortcut, though I can't test that on macOS. Once checked, tab can be used to reached all thee input fields on that screen.

I don't see any way to address "Needs to not add a QSettings prune option" other than #12833 or #13029, so unless there is a better idea out there this PR will likely be stuck for a while.

@jonasschnelli:
> I also think it should be labeled as "target" (something we want to reach but not sure we can).

Bit of a cop-out, but I think the tooltip "Actual disk usage may be somewhat higher" encapsulates that.

harding commented:
Lightly tested cbede7dbfde83d53ef38d257e9940af5f163b03c (but did not review code).  Updated tooltip and tooltip range LGTM, thanks!

mess110 commented:
Tested ACK cbede7d

ken2812221 commented on src/qt/optionsdialog.cpp:
```diff
@@ -266,6 +280,11 @@ void OptionsDialog::on_hideTrayIcon_stateChanged(int fState)
 
+void OptionsDialog::togglePruneWarning(bool enabled)
+{
+    ui->pruneWarning->setVisible(!ui->pruneWarning->isVisible());
```
`enabled` is unused

ken2812221 commented:
Concept ACK

laanwj commented:
utACK cbede7dbfde83d53ef38d257e9940af5f163b03c

luke-jr commented:
Why was this merged with QSettings?

Sjors commented:
#12833 gets rid of `QSettings` almost completely. Adding this one setting to the migration there is trivial

If #11082 makes it into 0.17 then I can remove the `QSettings` stuff in #12833 as if it was never there. 

 If it takes longer, then the only gotcha is that users need to put `prune=1` in `bitcoin.conf` if you want to use `bitcoind`.

jonasschnelli commented:
Post merge tested ACK
